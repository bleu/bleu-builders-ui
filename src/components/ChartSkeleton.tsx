import React from "react";
import { Skeleton } from "./ui/Skeleton";

function BarSet() {
  return (
    <>
      <Skeleton className="h-48 w-10" />
      <Skeleton className="h-64 w-10" />
      <Skeleton className="h-56 w-10" />
      <Skeleton className="h-64 w-10" />
    </>
  );
}
export function ChartSkeleton({ barsSetCount = 3, className = "" }) {
  const barSet = new Array(barsSetCount).fill(null);

  return (
    <div className={`${className} p-4`}>
      <div className="flex flex-col items-center">
        <div className="flex items-end space-x-2">
          <span className="sr-only">Loading...</span>
          {barSet.map((_, barIndex) => (
            // eslint-disable-next-line react/no-array-index-key
            <BarSet key={barIndex} />
          ))}
        </div>
      </div>
    </div>
  );
}
