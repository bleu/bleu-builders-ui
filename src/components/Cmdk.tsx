import { CircleIcon, FileIcon } from "@radix-ui/react-icons";
import * as React from "react";
import { useNavigate } from "react-router-dom";
import { Trans, useTranslation } from "react-i18next";
import { CommandLoading } from "cmdk";
import useSWR from "swr";
import {
  Button,
  ButtonProps,
  CommandDialog,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator,
} from "#/components/ui";
import { cn } from "#/lib/utils";
import { useDebounceValue } from "#/hooks/useDebounceValue";

interface Command {
  href: string;
  id?: string;
  result_type?: string;
  title: string;
  type?: string;
}

interface CommandMenuProps {
  commands: {
    mainNav: Command[];
    sidebarNav?: { items: Command[]; title: string }[];
  };
  fetcher: (query: string) => Promise<Command[]>;
  icons?: Record<string, React.ComponentType<{ className: string }>>;
}

export function CommandMenu({
  commands,
  fetcher,
  className,
  icons,
  ...props
}: CommandMenuProps & ButtonProps) {
  const navigate = useNavigate();
  const [open, setOpen] = React.useState(false);
  const [search, setSearch] = React.useState("");
  const [debouncedSearch, setDebouncedSearch] = useDebounceValue("", 300);
  const { data: searchResults, isLoading: loading } = useSWR<Command[]>(
    debouncedSearch,
    fetcher
  );

  React.useEffect(() => {
    setDebouncedSearch(search);
  }, [search]);

  const { t } = useTranslation();

  React.useEffect(() => {
    const down = (e: KeyboardEvent) => {
      if (e.key === "k" && (e.metaKey || e.ctrlKey)) {
        e.preventDefault();
        setOpen((prevOpen) => !prevOpen);
      }
    };

    document.addEventListener("keydown", down);
    return () => document.removeEventListener("keydown", down);
  }, []);

  const runCommand = (command: () => void) => {
    setOpen(false);
    command();
  };

  return (
    <>
      <Button
        variant="outline"
        className={cn(
          className,
          "text-muted-foreground relative w-full justify-start text-sm sm:pr-12 md:w-40 lg:w-64"
        )}
        onClick={() => setOpen(true)}
        {...props}
      >
        <span className="inline-flex">{t("Global search")}</span>
        <kbd className="bg-muted top- pointer-events-none absolute right-1.5 hidden h-5 select-none items-center gap-1 rounded border px-1.5 font-mono text-[10px] font-medium opacity-100 sm:flex">
          <span className="text-xs">âŒ˜</span>K
        </kbd>
      </Button>
      <CommandDialog open={open} onOpenChange={setOpen} loading={loading}>
        <CommandInput
          placeholder={t("Type a command or search")}
          value={search}
          onValueChange={setSearch}
        />
        <CommandList>
          <CommandEmpty>
            <Trans>No results found</Trans>.
          </CommandEmpty>

          {loading || searchResults !== undefined ? (
            <>
              {" "}
              <CommandGroup heading={t("Search results")}>
                {loading && (
                  <CommandLoading>
                    <p className="text-muted-foreground">
                      <Trans>Loading...</Trans>
                    </p>
                  </CommandLoading>
                )}{" "}
                {searchResults?.map((result) => {
                  const Icon =
                    icons && result.type ? icons[result.type] : FileIcon;

                  return (
                    <CommandItem
                      key={result.href}
                      value={result.title}
                      onSelect={() => {
                        runCommand(() => navigate(result.href));
                      }}
                    >
                      <p>
                        <Icon className="inline mr-1 h-2 w-2 stroke-1" />
                        <b className="inline">{result.result_type}</b> |{" "}
                        {result.title}
                      </p>
                    </CommandItem>
                  );
                })}
              </CommandGroup>
              <CommandSeparator />
            </>
          ) : (
            ""
          )}
          <CommandGroup heading="Links">
            {commands.mainNav.map((navItem) => (
              <CommandItem
                key={navItem.href}
                value={navItem.title}
                onSelect={() => {
                  runCommand(() => navigate(navItem.href));
                }}
              >
                <FileIcon className="mr-2 h-2 w-2" />
                {navItem.title}
              </CommandItem>
            ))}
          </CommandGroup>
          <CommandSeparator />
          {commands.sidebarNav?.map((group) => (
            <CommandGroup key={group.title} heading={group.title}>
              {group.items.map((navItem) => (
                <CommandItem
                  key={navItem.href}
                  value={navItem.title}
                  onSelect={() => {
                    runCommand(() => navigate(navItem.href));
                  }}
                >
                  <div className="mr-2 flex h-4 w-4 items-center justify-center">
                    <CircleIcon className="h-3 w-3" />
                  </div>
                  {navItem.title}
                </CommandItem>
              ))}
            </CommandGroup>
          ))}
        </CommandList>
      </CommandDialog>
    </>
  );
}
