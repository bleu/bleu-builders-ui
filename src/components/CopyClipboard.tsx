/* eslint-disable consistent-return */
import { ClipboardCopyIcon } from "@radix-ui/react-icons";
import copy from "copy-to-clipboard";
import React, { PropsWithChildren, useCallback, useLayoutEffect } from "react";
import { cn } from "#/lib/utils";
import { useToast } from "#/hooks/useToast";

export function ClickToCopy({
  text,
  className,
  children,
}: PropsWithChildren<{
  className?: string;
  text: string;
}>) {
  const [copied, setCopied] = React.useState(false);
  const { toast } = useToast();

  useLayoutEffect(() => {
    if (copied) {
      const timeout = setTimeout(() => {
        setCopied(false);
      }, 2000);

      toast({
        title: "Copied to clipboard",
        variant: "default",
      });

      return () => clearTimeout(timeout);
    }
  }, [copied]);

  return (
    <div>
      <CopyToClipboard
        text={text}
        onCopy={() => setCopied(true)}
        className={cn("flex flex-row items-center", className)}
      >
        <>
          <span className="mr-1">{children}</span>
          <ClipboardCopyIcon
            width={18}
            height={18}
            className="text-foreground"
          />
        </>
      </CopyToClipboard>
    </div>
  );
}

function CopyToClipboard({ text, className, onCopy, children }) {
  const [copied, setCopied] = React.useState(false);

  useLayoutEffect(() => {
    if (copied) {
      const timeout = setTimeout(() => {
        setCopied(false);
      }, 1000);

      return () => clearTimeout(timeout);
    }
  }, [copied]);

  const handleCopy = useCallback(() => {
    copy(text);
    onCopy?.();
  }, [text, onCopy]);

  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    <div className={cn("cursor-pointer", className)} onClick={handleCopy}>
      {children}
    </div>
  );
}
