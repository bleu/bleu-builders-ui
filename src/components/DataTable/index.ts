export * from "./DataTableColumnHeader";
export * from "./DataTableFacetedFilter";
export * from "./DataTablePagination";
export * from "./DataTableRowActions";
export * from "./DataTableToolbar";
export * from "./DataTableViewOptions";
export * from "./DynamicActionComponent";
export * from "./SWRDataTable";
export * from "./TableContext";
export * from "./useSWRDataTable";
export * from "./useTableState";
export * from "./DataTable";
export * from "./DataTableSearch";
export * from "./DataTableFilters";
