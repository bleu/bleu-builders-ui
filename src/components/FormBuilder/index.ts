export { fieldComponents } from "./builder";
export * from "./fields";
export * from "./types";
export * from "./buildForm";
