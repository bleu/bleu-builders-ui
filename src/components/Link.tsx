import { Link as RRDLink, type LinkProps } from "react-router-dom";
import React from "react";

export function Link({ to, children, ...props }: LinkProps) {
  return (
    <RRDLink to={to} unstable_viewTransition {...props}>
      {children}
    </RRDLink>
  );
}
