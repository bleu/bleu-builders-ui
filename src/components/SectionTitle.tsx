import React from "react";

type SectionTitleProps = {
  children: React.ReactNode;
};

export const SectionTitle: React.FC<SectionTitleProps> = ({ children }) => (
  <h2 className="pt-8 text-2xl font-bold tracking-tigh text-foreground">
    {children}
  </h2>
);
