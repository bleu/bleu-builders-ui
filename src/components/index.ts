export * from "./Cmdk";
export * from "./CopyClipboard";
export * from "./TaggablePopover";
export * from "./ThemeToggle";
export * from "./DataTable";
export * from "./FormBuilder";
export * from "./ui";
export * from "./RailsApp";
export * from "./SectionTitle";
export * from "./OtpInput";
export * from "./SubmitButton";
export * from "./StrictModeDroppable";
export * from "./Skeletons";
export * from "./Plot";
