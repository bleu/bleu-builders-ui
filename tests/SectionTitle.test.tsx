import renderer from "react-test-renderer";
import { expect, it } from "vitest";
import { SectionTitle } from "../src/components/SectionTitle";

it("renders correctly", () => {
  const tree = renderer
    .create(<SectionTitle>Section Title Content</SectionTitle>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
