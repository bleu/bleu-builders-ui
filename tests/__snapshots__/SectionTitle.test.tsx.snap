// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<h2
  className="pt-8 text-2xl font-bold tracking-tigh text-foreground"
>
  Clicked this many times
</h2>
`;

exports[`renders correctly w/o props 1`] = `
<h2
  className="pt-8 text-2xl font-bold tracking-tigh text-foreground"
>
  Clicked this many times
</h2>
`;
